### PR TITLE
ARROW-15 Add support for BSON Decimal128 type

### PR DESCRIPTION
**Pull request by blink1073**

### CODE REVIEW

File: bindings/python/pymongoarrow/lib.pyx [hunk 1]
```diff
@@ -59,11 +60,14 @@ def process_bson_stream(bson_stream, context):
     cdef bson_reader_t* stream_reader = bson_reader_new_from_data(docstream, length)
     cdef const bson_t * doc = NULL
     cdef bson_iter_t doc_iter
+    cdef bson_decimal128_t dec128
     cdef const char* key
     cdef bson_type_t value_t
     cdef Py_ssize_t count = 0
     cdef const char * bson_str
     cdef uint32_t str_len
+    cdef char *decimal128_str = <char *> malloc(
```
Comment (ShaneHarvey) [inline bindings/python/pymongoarrow/lib.pyx:69]: Does this need to be a malloc or can it reside on the stack?

File: bindings/python/pymongoarrow/lib.pyx [hunk 2]
```diff
@@ -115,6 +119,10 @@ def process_bson_stream(bson_stream, context):
                         if value_t == BSON_TYPE_UTF8:
                             bson_str = bson_iter_utf8(&doc_iter, &str_len)
                             builder.append(<bytes>(bson_str)[:str_len])
+                        elif value_t == BSON_TYPE_DECIMAL128:
+                            bson_iter_decimal128(&doc_iter, &dec128)
+                            bson_decimal128_to_string(&dec128, decimal128_str)
+                            builder.append(<bytes>(decimal128_str))
```
Comment (ShaneHarvey) [inline bindings/python/pymongoarrow/lib.pyx:125]: Does builder.append take a copy of the decimal128_str?

File: bindings/python/pymongoarrow/types.py [hunk 3]
```diff
@@ -45,6 +46,17 @@ def __reduce__(self):
         return ObjectIdType, ()
 
 
+class Decimal128StringType(PyExtensionType):
+    _type_marker = _BsonArrowTypes.decimal128_str
+
+    def __init__(self):
+        # Delayed to prevent circular import.
```
Comment (ShaneHarvey) [inline bindings/python/pymongoarrow/types.py:53]: What does "delayed" refer to?

File: bindings/python/test/test_bson.py [hunk 4]
```diff
@@ -163,3 +178,32 @@ def test_simple(self):
...
+    def test_simple(self):
+        docs = [
+            {"data": "a"},
+            {"data": "dätá"},
+        ]
+        as_dict = {"data": ["a", "dätá"]}
+        self._run_test(docs, as_dict)
+
+
+class TestDecimal128StringType(TestBsonToArrowConversionBase):
```
Comment (ShaneHarvey) [inline bindings/python/test/test_bson.py:197]: We also need to test the pandas/numpy conversion by testing Decimal128 with find_arrow/pandas/numpy

**Review by ShaneHarvey — CHANGES_REQUESTED** [review]

File: bindings/python/pymongoarrow/lib.pyx (hunk 1 above)
Comment (blink1073) [reply bindings/python/pymongoarrow/lib.pyx:69]: It has to be a malloc unless we give it a literal number for the length.

File: bindings/python/pymongoarrow/lib.pyx (hunk 2 above)
Comment (blink1073) [reply bindings/python/pymongoarrow/lib.pyx:125]: I don't follow, you mean copy vs by reference?  It is coerced to a Python object either way, I was just making it explicit.

File: bindings/python/pymongoarrow/types.py (hunk 3 above)
Comment (blink1073) [reply bindings/python/pymongoarrow/types.py:53]: Stale comment, removed

File: bindings/python/test/test_bson.py (hunk 4 above)
Comment (blink1073) [reply bindings/python/test/test_bson.py:197]: Done

File: bindings/python/pymongoarrow/lib.pyx (hunk 1 above)
Comment (ShaneHarvey) [reply bindings/python/pymongoarrow/lib.pyx:69]: BSON_DECIMAL128_STRING is a literal number though. See http://mongoc.org/libbson/current/bson_decimal128_t.html#example

Can we do:
```
cdef char decimal128_str[BSON_DECIMAL128_STRING];
```

File: bindings/python/pymongoarrow/lib.pyx (hunk 2 above)
Comment (ShaneHarvey) [reply bindings/python/pymongoarrow/lib.pyx:125]: SGTM

File: bindings/python/pymongoarrow/lib.pyx (hunk 1 above)
Comment (blink1073) [reply bindings/python/pymongoarrow/lib.pyx:69]: I had tried that originally:

```
cdef char decimal128_str[BSON_DECIMAL128_STRING];
                                  ^
      ------------------------------------------------------------

      pymongoarrow/lib.pyx:69:29: Not allowed in a constant expression
```

Comment (ShaneHarvey) [reply bindings/python/pymongoarrow/lib.pyx:69]: Oh, that's odd. One last question, does using a compile time redefinition work? EG:
```
DEF DEC128_STR_LEN = BSON_DECIMAL128_STRING
...
    cdef char decimal128_str[BSON_DECIMAL128_STRING];
```

https://cython.readthedocs.io/en/latest/src/userguide/language_basics.html#compile-time-definitions

**Review by ShaneHarvey — APPROVED** [review]
LGTM assuming the `DEF` approach doesn't work either.

**Comment by blink1073** [comment]
No luck:

```
  DEF DEC128_STR_LEN = BSON_DECIMAL128_STRING
                      ^
  ------------------------------------------------------------

  pymongoarrow/lib.pyx:49:21: Compile-time name 'BSON_DECIMAL128_STRING' not defined
```

**Pull request merged**